configure.ml

```diff
@@ -1391,3 +1391,23 @@ let write_configpy f =
   pr_s "version" coq_version
 
 let _ = write_configpy "config/coq_config.py"
+
+let copy_text_file src tgt =
+  let i = open_in src in
+  let o = open_out tgt in
+  try
+    while true do
+      let s = input_line i in
+      output_string o s;
+      output_char o '\n'
+    done
+  with End_of_file ->
+    close_in i;
+    close_out o
+
+let _ =
```

Comment:
I am not sure what you are trying to do, but certainly configure is not the place to do this. Move to CoqIDE makefile, or even better make CoqIDE smarter so it can choose the file.

Reply:
I need the platform-dependent version of the file `ide/coqide.keys` to exist also in `-local` mode, this is why I put it in `configure`. This file has to be "generated" before any form of `make install` because `-local` does not install. The `cp` could be done in, say, `make  bin/coqide`, but e.g. `dev/ocamldebug-coq` is generated by `configure.ml` and not by `make`, so I did the same.

Reply:
First, no reason why you can't have coqide select the right file based on the platform and you remove a problem. Modify `preferences.ml` so the name is selected in base of the platform.

Would you find a reason not to do that [bad idea], then `coqide.keys` would be an auto-generated file depending on other sources files, thus `configure` is the wrong place to generate it as it won't track the dependencies correctly.

Reply:
> First, no reason why you can't have coqide select the right file based on the platform and you remove a problem. 

What do you mean? That we should install all platform-dependent variants of the default `coqide.keys` template whatever the installation target is? (So that `coqide`, is dynamically able to select the default template based on the platform.)

> Would you find a reason not to do that [bad idea], then coqide.keys would be an auto-generated file depending on other sources files, thus configure is the wrong place to generate it as it won't track the dependencies correctly.

I don't understand which other dependencies we would have to track. I'm talking here about the template which serves to generate the `coqide.keys` file the first time `coqide` is used.

Reply:
> What do you mean? That we should install all platform-dependent variants of the default `coqide.keys` template whatever the installation target is? (So that `coqide`, is dynamically able to select the default template based on the platform.)

I do mean that indeed.

> I don't understand which other dependencies we would have to track. I'm talking here about the template which serves to generate the `coqide.keys` file the first time `coqide` is used.

You need to track the source file. For example when cross building, or if you update the source file the copy will become stale unless you call configure again.

Reply:
Configure is not a build system, it's purpose is to query the few things the build system cannot and then let the build system proceed. In the case of Coq, it means that we can get rid of configure altogether [and we hope we will].

Reply:
> I do mean that indeed.

OK, I can do that, but this would be inconsistent, as I said, with e.g. the way `dev/ocamldebug-coq` is built (which indeed is not updated automatically if the `configure.ml` is updated). Also, the current way is not impeding cross building: If you want to build for platform X on platform Y, you will tell the configure to prepare the `coqide.keys` appropriate for platform X.

Then, if you have in mind a view where `configure` disappears, that's another story. Can you tell more? Is that a standard enough approach in the ecosystem that users would not get disoriented?

Reply:
I'd say that no configure is fairly standard these days, anyways users are not mean to see configure even today.

Building-wise remembering when to run configure is generally a pain, anyways, this kind of files are better suited for make.

Reply:
@ejgallego: I have nothing against your proposal but I remain skeptical that what I did is non standard or "bad" per se. For instance, we install the MacOS specific stuff (like `coqide.icns`) only on MacOS. Do you mean that we could want, say, to install coq on Linux but use this installation to run CoqIDE in a MacOS image and thus need the MacOS file even in the Linux installation? Otherwise said, why not doing what you suggest systematically then, for all platform-dependent files?

If the point is to get rid of `configure` why not to do all these getting-rid-of-configure at the same time, but in the meantime preserving some consistency.

So, shortly, I feel a bit lost in how imperative your request is regarding a future which you seem to see clearly but which I have some difficulty myself to foresee. Why should I change the way coqide is looking for its files just to avoid running `configure` which we anyway need to run at the current time, and which I don't see how we could avoid having to run.

Reply:
I didn't mean such esoteric stuff, it was just a suggestion. It is OK if you want to add more complexity to the build system instead of the particular application is affected, but auto-generated files should be handled by make not by configure. Just add make rule for the file and dependencies will be properly tracked.

In the end, what matters is where do you push the complexity of maintaining consistency of the build, and what is the effort of doing so.

Reply:
My request about this not going in configure is pretty strong. Regarding using `make` vs having CoqIDE load the proper file, I do understand that using OCaml to implement the second is far easier than having to write fairly tricky make scripts, however there I shall remain neutral if you feel writing make is easier than OCaml.

Reply:
Also, why is configure a problem? Because you never know when you have to re-run it. In fact, in Dune, running configure is not needed as in fact the files that are produced are handled in the standard dependency chain. So you just do `dune build` and forget about what has to be done.

Reply:
A good example of how painful configure is the problems we have when changing OCaml compiler switch. If you change your compiler and forget to do configure you get weird problems; this is better managed by the build system which can properly track a dependency on the compiler.

Reply:
So this is not about "removing" configure so much, but moving its functionality to a system that manages it automatically.

Reply:
> I do understand that using OCaml to implement the second is far easier than having to write fairly tricky make scripts, however there I shall remain neutral if you feel writing make is easier than OCaml.

I'm very sorry @ejgallego, I don't feel at writing fairly tricky scripts now that I have a version which works and which seems compatible with the spirit of what exists in configure.ml currently. If you know exactly what you want and how to implement it (and incidentally how to make it compatible with dune), would you be ok to force-push what you want to this branch?

Reply:
I am afraid that I don't have the time now to understand what this file is supposed to do; I am afraid that I cannot ACK the addition you propose to configure.

Reply:
Looks like a deadlock!

Reply:
An additional and stronger reason is that up to this point `ide` is fairly independent [that is, it can build in isolation, which is important for packages]; after you patch, such property would be lost.

Reply:
How much time would you need to remove `configure.ml`, or at least to move the building of `dev/ocamldebug-coq` out of `configure.ml`?

Reply:
Maybe a third opinion could help me to understand what is the right direction to go? Is there something wrong with my approach, considering the current situation? What would e.g. a nix expert do?